File: src/camera.rs
use anyhow::Context;
use gio::{prelude::*, subclass::prelude::*};
use glib::{clone, subclass::Signal};
use gst::prelude::*;
use once_cell::unsync::OnceCell;

mod imp {
    use super::*;

    #[derive(Debug, Default)]
    pub struct Camera {
        pub pipeline: OnceCell<gst::Pipeline>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for Camera {
        const NAME: &'static str = "BthCamera";
        type Type = super::Camera;
        type ParentType = glib::Object;
    }

    impl ObjectImpl for Camera {
        fn signals() -> &'static [Signal] {
            use once_cell::sync::Lazy;
            static SIGNALS: Lazy<Vec<Signal>> = Lazy::new(|| {
                vec![Signal::builder(
                    "code-detected",
                    &[String::static_type().into()],
                    <()>::static_type().into(),
                )
                .build()]
            });
            SIGNALS.as_ref()
        }

        fn constructed(&self, obj: &Self::Type) {
            self.parent_constructed(obj);

            if let Err(err) = obj.setup_pipeline() {
                log::error!("Failed to setup pipeline: {:#}", err);
            }
        }

        fn dispose(&self, _obj: &Self::Type) {
            self.pipeline
                .get()
                .unwrap()
                .set_state(gst::State::Null)
                .unwrap();
        }
    }
}

glib::wrapper! {
    pub struct Camera(ObjectSubclass<imp::Camera>);
}

impl Camera {
    pub fn new() -> Self {
        glib::Object::new(&[]).expect("Failed to create Camera.")
    }

    pub fn connect_code_detected<F>(&self, f: F) -> glib::SignalHandlerId
    where
        F: Fn(&Self, &str) + 'static,
    {
        self.connect_local("code-detected", true, move |values| {
            let obj = values[0].get::<Self>().unwrap();
            let code = values[1].get::<String>().unwrap();
            f(&obj, &code);
            None
        })
        .unwrap()
    }

    pub fn start(&self) -> anyhow::Result<()> {
        let imp = imp::Camera::from_instance(self);
        let pipeline = imp.pipeline.get().unwrap();

        let bus = pipeline.bus().unwrap();
        bus.add_watch_local(
            clone!(@weak self as obj => @default-return Continue(false), move |_, message| {
                obj.handle_bus_message(message)
            }),
        )
        .unwrap();

        pipeline
            .set_state(gst::State::Playing)
            .context("Failed to set pipeline state to Playing")?;

        Ok(())
    }

    pub fn stop(&self) -> anyhow::Result<()> {
        let imp = imp::Camera::from_instance(self);
        let pipeline = imp.pipeline.get().unwrap();

        pipeline.set_state(gst::State::Null)?;
        let bus = pipeline.bus().unwrap();
        bus.remove_watch()?;

        Ok(())
    }

    fn setup_pipeline(&self) -> anyhow::Result<()> {
        let pipeline = gst::Pipeline::new(None);

        let pipewiresrc = gst::ElementFactory::make("pipewiresrc", None)?;
        let queue = gst::ElementFactory::make("queue", None)?;
        let videoconvert = gst::ElementFactory::make("videoconvert", None)?;
        let zbar = gst::ElementFactory::make("zbar", None)?;
        let fakesink = gst::ElementFactory::make("fakesink", None)?;

        let elements = &[&pipewiresrc, &queue, &videoconvert, &zbar, &fakesink];
        pipeline.add_many(elements)?;
        gst::Element::link_many(elements)?;

        for e in elements {
            e.sync_state_with_parent()?;
        }

        let imp = imp::Camera::from_instance(self);
        imp.pipeline.set(pipeline).unwrap();

        Ok(())
    }

    fn handle_bus_message(&self, message: &gst::Message) -> Continue {
        use gst::MessageView;

        match message.view() {
            MessageView::Error(err) => {
                log::error!(
                    "Error from {:?}: {} ({:?})",
                    err.src().map(|s| s.path_string()),
                    err.error(),
                    err.debug()
                );

                self.stop().unwrap();

                Continue(false)
            }
            MessageView::StateChanged(sc) => {
                let imp = imp::Camera::from_instance(self);
                let pipeline = imp.pipeline.get().unwrap();

                if message.src().as_ref() == Some(pipeline.upcast_ref::<gst::Object>()) {
                    log::info!(
                        "Pipeline state set from {:?} -> {:?}",
                        sc.old(),
                        sc.current()
                    );
                }
                Continue(true)
            }
            MessageView::Element(e) => {
                let symbol: String = e.structure().unwrap().get("symbol").unwrap();

                self.emit_by_name("code-detected", &[&symbol]).unwrap();

                Continue(true)
            }
            _ => Continue(true),
        }
    }
}

impl Default for Camera {
    fn default() -> Self {
        Self::new()
    }
}

File: src/camera.py
from gi.repository import GObject, Gst, GLib


class ElementNotFoundError(Exception):
    def __init__(self, element_name: str):
        self._element_name = element_name

    def __str__(self):
        return f"Element {self._element_name} not found"


class Camera(GObject.Object):
    __gsignals__ = {"code-detected": (GObject.SIGNAL_RUN_LAST, None, (str,))}

    _pipeline: Gst.Pipeline
    _bus: Gst.Bus

    def __init__(self):
        super().__init__()

        self._pipeline = Gst.Pipeline()
        self._bus = self._pipeline.get_bus()

        try:
            pipewiresrc = make_gst_element("pipewiresrc")
            queue = make_gst_element("queue")
            videoconvert = make_gst_element("videoconvert")
            zbar = make_gst_element("zbar")
            fakesink = make_gst_element("fakesink")
        except ElementNotFoundError as error:
            print(f"Error: {error}")

        elements = [pipewiresrc, queue, videoconvert, zbar, fakesink]

        for element in elements:
            self._pipeline.add(element)

        pipewiresrc.link(queue)
        queue.link(videoconvert)
        videoconvert.link(zbar)
        zbar.link(fakesink)

        for element in elements:
            element.sync_state_with_parent()

    def start(self):
        self._bus.add_signal_watch()
        self._bus_handler_id = self._bus.connect("message", self._handle_message)
        self._pipeline.set_state(Gst.State.PLAYING)
        print(">>> Camera started")

    def stop(self):
        self._pipeline.set_state(Gst.State.NULL)
        self._bus.disconnect(self._bus_handler_id)
        self._bus.remove_signal_watch()

    def _handle_message(self, bus: Gst.Bus, message: Gst.Message):
        if message.type == Gst.MessageType.ELEMENT:
            symbol = message.get_structure().get_value("symbol")
            if symbol is not None:
                self.emit("code-detected", symbol)
            return True

        if message.type == Gst.MessageType.STATE_CHANGED:
            if message.src == self._pipeline:
                old_state, new_state, pending = message.parse_state_changed()
                print(f">>> Pipeline state set from {old_state} -> {new_state}")
            return True

        if message.type == Gst.MessageType.ERROR:
            error, debug = message.parse_error()
            print(f"Error: {error} ({debug})")
            self.stop()
            return False


def make_gst_element(name: str) -> Gst.Element:
    element = Gst.ElementFactory.make(name)

    if element is None:
        raise ElementNotFoundError(name)

    return element
